File: cpp/trees/0116-populating-next-right-pointers-in-each-node.cpp
#include <queue>
#include <structures.h>

using namespace std;


Node* connect(Node* root) {
    if (root == nullptr) return root;
    
    queue<Node*> q;
    q.push(root);
    
    Node* prev = root;
    
    while (!q.empty()) {
        int level = q.size();
        for (int i = 0; i < level; i++) {
            Node* cur = q.front();
            q.pop();
            
            if (i > 0) {
                prev->next = cur;
            }
            prev = cur;
            
            if (cur->left) q.push(cur->left);
            if (cur->right) q.push(cur->right);
        }
    }
    
    return root;
}

File: js/trees/0116-populating-next-right-pointers-in-each-node.js
var connect = function(root) {
    if (root === null) return null;
    var queue = [root];
    
    while(queue.length > 0) {
        lvl_sz = queue.length;
        
        var prev = null;
        for (var i = 0; i < lvl_sz; i++) {
            var cur = queue.shift();
            
            if (i !== 0 && prev !== null) {
                prev.next = cur;
            }
            prev = cur;
            
            if (cur !== null) queue.push(cur.left);
            if (cur !== null) queue.push(cur.right);
        }
    }
    
    return root;
};